internal/tiltfile/tiltfile_state.go

```diff
@@ -616,6 +616,49 @@ func (s *tiltfileState) assembleK8sV2() error {
 				r.name = opts.newName
 				s.k8sByName[r.name] = r
 			}
+
+			selectors := make([]k8sObjectSelector, len(opts.objects))
+			for i, o := range opts.objects {
+				s, err := selectorFromString(o)
+				if err != nil {
+					return errors.Wrapf(err, "Error making selector from string '%s'", o)
+				}
+				selectors[i] = s
+			}
+
+			for i, selector := range selectors {
+				if len(s.k8sUnresourced) == 0 {
+					return fmt.Errorf("Unable to find any unresourced matches for selector %s. All YAML has already been resourced", opts.objects[i])
+				}
+				matches := []k8s.K8sEntity{}
+				for _, e := range s.k8sUnresourced {
+					if selector.matches(e) {
+						matches = append(matches, e)
+					}
+				}
+
+				for _, k := range s.k8s {
+					for _, e := range k.entities {
+						if selector.matches(e) {
+							matches = append(matches, e)
+						}
+					}
+				}
+				numMatches := len(matches)
+				if numMatches == 0 {
+					namesOfAvailableUnresourcedYAML := make([]string, len(s.k8sUnresourced))
+					for i, e := range s.k8sUnresourced {
+						namesOfAvailableUnresourcedYAML[i] = selectorStringFromK8sEntity(e)
+					}
+					return fmt.Errorf("Unable to find any unresourced matches for selector %s. Available unresourced YAML: %s", opts.objects[i], strings.Join(namesOfAvailableUnresourcedYAML, ", "))
+				}
+				if numMatches > 1 {
+					return fmt.Errorf("Found multiple (%d) matches for selector %+v: %v", numMatches, selector, matches)
+				}
+
+				e := matches[0]
+				s.addEntityToResourceAndRemoveFromUnresourced(e, r)
```

Comment:
If I'm reading this correctly, we'll end up panicking if the one match is from s.k8s rather than s.k8sUnresourced. I think tests for these cases would be useful:
* one resource specifies another resource's workload
* two resources specify the same object

(I think TestK8sResourceObjectsCantIncludeSameObjectTwice will also panic if you add another object to the unresourced yaml - it's only avoiding the panic because it's hitting the `if len(s.k8sUnresourced) == 0 {` check)

something like this might be more straightforward (or maybe it only feels more straightforward because pseudocode doesn't have to deal with go's verbosity?)
```
// figure out what resources claim what objects, validating constraints as we go
for r in resources:
  for sel in r.selectors:
    resourcedMatches = filter(resourcedObjects, sel)
    if len(resourcedMatches) > 0:
      error
    unresourcedMatches = filter(resourcedObjects, sel)
    if len(resourcedMatches) != 1:
      error
    if claims[unresourcedMatches[0]]:
      error // can print claims[unresourcedMatches[0]] and r+sel as the conflicting claims on the object
    claims[unresourcedMatches[0]] = {Resource: r, Selector: sel}

// at this point, we've matched all selectors and know we followed the rules, so just assign objects to resources
for obj, claim in claims:
  // assign obj to claim.Resource
```

Reply:
FWIW I wrote up [this matrix of possibilities](https://docs.google.com/spreadsheets/d/1arxzEk_83LI7F4W8NFSL2mbmNiyho0xUrrT5TWeciOg/edit#gid=0): does that seem right?

Reply:
FYI there's already a test case for two resources that specify the same workload, that's handled elsewhere.

I just added a test case for two resources specifying the same object.

Reply:
> (I think TestK8sResourceObjectsCantIncludeSameObjectTwice will also panic if you add another object to the unresourced yaml - it's only avoiding the panic because it's hitting the if len(s.k8sUnresourced) == 0 { check)

Yep good call